### DISCUSSION TITLE
select(x.assume_not_null()) - for when query has filter(x.is_not_null())

### DISCUSSION THREAD

**Discussion by Ten0** [diesel-rs/diesel#2819] [Ideas]
A kind of query I regularly encounter and gets pretty verbose is:
```rust
let mappings = xxx::table
    .select((xxx::id, xxx::some_property))
    .filter(xxx::some_property.is_not_null())
    .load::<(i32, Option<i32>)>(db)?
    .into_iter()
    .map(|(id, some_property)| Ok((id, some_property.ok_or_else(|| err_msg("queried not null"))))
    .collect::<Result<_,_>>()?;
```

If we had a way in the query builder to assert that as per the way we constructed the query, some fields can't be null (simple `Expression` that goes from nullable to non nullable (the opposite of `.nullable()`)), we could make that way less verbose:

```rust
let mappings = xxx::table
    .select((xxx::id, xxx::some_property.assume_not_null()))
    .filter(xxx::some_property.is_not_null())
    .load::<(i32, i32)>(db)?;
```

`load` would return an `UnexpectedNull` deserialization error should the `assume` be incorrect, which would be a reasonable error message (except in the edge-case of the `Option` deserialization which catches that error, in which case it may have kind of a silent-fail-ish behavior if the `assume` was incorrect - this could probably be solved by introducing an `AssumeNotNull<ST>` sql type kind of like `Nullable<ST>` and making a few error conversions here and there, but I'm not sure the added complexity would be worth it).

Does this seem reasonable?

**Comment by weiznich** [comment]
Having something like this is completely reasonable. As this is something that can be useful for a general audience I would be fine to include something like this into diesel, as long as the documentation clearly states that this can result in runtime errors.

**Comment by Ten0** [comment]
https://github.com/diesel-rs/diesel/pull/2820
